### PR TITLE
Add option group file_type

### PR DESCRIPTION
**Pull request by dontub**
Overview
----------------------------------------
This PR adds an option group for the `file_type_id` field of the `File` entity. (As far as the comment of field indicates there already should be such an option group.)

Before
----------------------------------------
`file_type_id` is just an integer.

After
----------------------------------------
`file_type_id` is an integer referring to the option group `file_type`.

Technical Details
----------------------------------------
The file `sql/civicrm_generated.mysql` is not changed. I tried to regenerate it though the resulting file seem to contain the same data, but in different order. Thus the diff was really large and I decided to skip that for now. Is there a description on how to do this correctly?

If you in general agree with this new option group it has to be ensured that it is added on update.

### CODE REVIEW

**Comment by demeritcowboy** [comment]
I think you do need to regenerate civicrm_generated since this data is in there. There's some docs at https://docs.civicrm.org/dev/en/latest/framework/database/#binregensh

**Comment by dontub** [comment]
I've updated `civicrm_generated.sql` using the GitHub action. Though the diff is nevertheless really big.

With regard to the upgrade: As far as I can see there are two approaches:

Either add
```php
CRM_Core_BAO_OptionGroup::ensureOptionGroupExists([
      'name' => 'file_type',
      'title' => ts('File Type'),
      'data_type' => 'Integer',
      'is_reserved' => 1,
    ]);
```
to a file in `CRM/Upgrade/Incremental/php/`

or add
```sql
INSERT INTO `civicrm_option_group` (`name`, {localize field='title'}`title`{/localize}, `data_type`, `is_reserved`)
VALUES ('file_type' , {localize}'{ts escape="sql"}File Type{/ts}'{/localize}, 'Integer', 1);
```
to a file in `CRM/Upgrade/Incremental/sql/`.

Which approach should I choose and which one is the correct target version?

**Comment by colemanw** [comment]
@dontub you should use `ensureOptionGroupExists` and place it in an upgrade task called from `function upgrade_5_61_alpha1`

**Comment by dontub** [comment]
Upgrade task is added.

File: CRM/Upgrade/Incremental/php/FiveSixtyOne.php [hunk 1]
```diff
@@ -63,6 +63,16 @@ public function upgrade_5_61_alpha1($rev): void {
...
     $this->addTask(ts('Update Saved Mapping for contribution import', [1 => $rev]), 'convertMappingFieldsToApi4StyleNames', $rev);
+    $this->addTask(
+      'Add option group for file_type_id in file table',
+      'addOptionGroup',
+      [
+        'name' => 'file_type',
+        'title' => ts('File Type'),
+        'data_type' => 'Integer',
+        'is_reserved' => 1,
+      ]
```
Comment (colemanw) [inline CRM/Upgrade/Incremental/php/FiveSixtyOne.php:74]: Test failure is:
```
Too few arguments to function CRM_Upgrade_Incremental_Base::addOptionGroup(  
  ), 2 passed in /home/jenkins/bknix-dfl/build/build-0/web/sites/all/modules/  
  civicrm/CRM/Queue/Task.php on line 101 and exactly 3 expected
```

Looks like you need to add another argument here which is an array of option values.

**Comment by colemanw** [comment]
@dontub it looks like the only thing missing is the option values, both in the installer and the upgrader.

**Comment by dontub** [comment]
> @dontub it looks like the only thing missing is the option values, both in the installer and the upgrader.

Which file types should be available by default?

**Comment by colemanw** [comment]
> Which file types should be available by default?

Umm, I probably should have asked this earlier, but... what is this field actually used for? Is it shown anywhere in the UI? If so, what options are shown in the dropdown for this field?

**Comment by dontub** [comment]
> > Which file types should be available by default?
> 
> Umm, I probably should have asked this earlier, but... what is this field actually used for? Is it shown anywhere in the UI? If so, what options are shown in the dropdown for this field?

This is what the field's comment says: https://github.com/civicrm/civicrm-core/blob/fe4d8c240eb6de41fffee53f29c5c02c522c8cf3/xml/schema/Core/File.xml#L29

It doesn't seem to be used by Core, yet.

We aim to use it in an extension. The file types we want to use are extension specific. So we'd add the required values to the option group via managed entities.

**Comment by dontub** [comment]
> @dontub in that case, perhaps it's sufficient to just pass an empty array as the last argument to `addOptionGroup`.

Done.

**Comment by colemanw** [comment]
Ok this looks done. We just need to move the upgrade function over to `CRM_Upgrade_Incremental_php_FiveSixtyTwo`

**Comment by colemanw** [comment]
@dontub if CiviCRM adds to this list [as I'm proposing here](https://lab.civicrm.org/dev/core/-/issues/4251) we'll start with `1`, `2`, `3`, etc. as the option value numbers so to prevent conflicts your extension should follow the best-practice of *not* specifying `value` when it creates option values and allowing it to auto-increment.

**Comment by dontub** [comment]
@colemanw Inserting an `OptionValue` as managed entity without value didn't work (has that changed?). So our solution was to use some high number. (Of course that's not ideal.) Actually I'd had preferred it to reference the file type via string instead of an integer. (This holds true for other fields referencing an `OptionValue`, e.g. activity type.)

**Comment by colemanw** [comment]
> Actually I'd had preferred it to reference the file type via string instead of an integer. (This holds true for other fields referencing an OptionValue, e.g. activity type.)

I agree, but unfortunately this design decision was made a long time ago and cannot be easily undone.

**Comment by colemanw** [comment]
> Inserting an OptionValue as managed entity without value didn't work (has that changed?).

Did you try it with an APIv4 managed entity? That might work.

**Comment by dontub** [comment]
> > Inserting an OptionValue as managed entity without value didn't work (has that changed?).
> 
> Did you try it with an APIv4 managed entity? That might work.

I tried it again and it worked now. Though in previous versions it failed.

**Pull request merged**